### PR TITLE
[LMS-347][Integration] Sign In page

### PR DESCRIPTION
**Pull request by ragg09-sunasterisk**
## Issue Link
[LMS-347 [Integration] Sign-in Page](https://framgiaph.backlog.com/view/LMS-347)

## Defintion of Done
* [x] it must be able to sign-in
* [x] it must redirect to home page after sign-in `/home`
* [x] when signed-in, it should not be able to access landing page `/`. When input manually, it should auto redirect to `/home`
* [x] when signed-in, it should not be able to access sign in page `/auth/sign-in`. When input manually, it should auto redirect to `/home`
* [x] It must display dynamic user or profile name
* [x] it must retrieve the `token` from backend and store it to cache or localstorage to keep the user signed-in when accessing to different page
* [x] also store necessary user details
* [x] create auth middleware

## Steps to reproduce
Follow the step to reproduce of the new migration file:
[Reference](https://github.com/framgia/sph-lms/pull/55)

Got to [Sign In Page](http://localhost:3000/auth/sign-in)
 - Creds
 -- admin@gmail.com
 -- password
 - access landing page manually (should return to /home)
 - access sign in page manually (should return to /home)
 
## Affected Components / Functionalities / Page
n/a

## Test Cases
_will update soon..._

## Notes
- /home page has hydration error which is not a part of my task, just close the error modal adn test features of my task only

## Screenshots
image
image
image
image

### CODE REVIEW

File: client/src/apis/index.ts [hunk 1]
```diff
@@ -1,7 +1,12 @@
 import axios from 'axios';
+import { getUserToken } from '../shared/utils';
```
Comment (kdacudag) [inline client/src/apis/index.ts:2]: Use module path aliases instead. Apply only to all modified lines authored by you. You can search for '../'

```suggestion
import { getUserFullName, isSignedIn } from '@/src/shared/utils';
```

File: client/src/shared/components/Navbar/index.tsx [hunk 2]
```diff
@@ -37,36 +38,32 @@ const Navbar: React.FC<NavbarProps> = ({ navItems, dropdownItems }) => {
...
             </div>
           </div>
-          {isSignedIn()
-            ? (
+          {isSignedIn() ? (
             <div className="flex items-center">
               <Link href={'/settings'} className="text-gray-600">
                 <SettingsIcon width={20} height={20} className="mr-2" />
               </Link>
               <Avatar name="John Doe" />
```
Comment (kdacudag) [inline client/src/shared/components/Navbar/index.tsx:63]: Add `getUserFullName()` to Avatar's name props.

File: api/app_sph_lms/api/views.py [hunk 3]
```diff
@@ -35,6 +35,7 @@ def get_auth_user(request):
         user = request.user
         token = Token.objects.get(user=user)
         return Response({
+            'full_name': user.first_name + ' ' + user.last_name,
```
Comment (escbooster12-dev) [inline api/app_sph_lms/api/views.py:38]: - you don't need to add `full_name` attribute all the time.
- use https://www.django-rest-framework.org/api-guide/fields/#serializermethodfield

File: client/src/pages/_app.tsx [hunk 4]
```diff
@@ -3,11 +3,13 @@ import '@/styles/globals.css';
...
 import 'react-toastify/dist/ReactToastify.css';
 import { ToastContainer } from 'react-toastify';
+import { useAuthMiddleware } from '../shared/hooks/useAuthMiddleware';
 
 const App: React.FunctionComponent<AppProps> = ({
   Component,
   pageProps
 }: AppProps) => {
+  useAuthMiddleware();
   return (
```
Comment (escbooster12-dev) [inline client/src/pages/_app.tsx:13]:
```suggestion
  useAuthMiddleware();

  return (
```

File: client/src/shared/utils/auth.ts [hunk 5]
```diff
@@ -1,3 +1,24 @@
...
+  if (
+    typeof localStorage !== 'undefined' &&
+    localStorage.getItem('user_token') !== undefined
+  ) {
+    return localStorage.getItem('user_token');
+  } else {
+    return null;
+  }
+};
+export const getUserFullName = (): string | null => {
```
Comment (escbooster12-dev) [inline client/src/shared/utils/auth.ts:15]:
```suggestion
};

export const getUserFullName = (): string | null => {
```

File: client/src/shared/hooks/useAuthMiddleware.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+/* eslint-disable @typescript-eslint/no-floating-promises */
+import axiosInstance from '@/src/apis';
```
Comment (escbooster12-dev) [inline client/src/shared/hooks/useAuthMiddleware.ts:2]:
```suggestion
import API from '@/src/apis';
```

**Review by Marielleregner7 — APPROVED** [review]
**APPROVED**
- [x] Able to sign-in.
- [x] redirected to home page after sign-in /home
- [x] after signing in, can no longer access landing page.
- [x] No ESLint warnings or errors.

Comments: 6
Test time: 45 mins
Iterations: 1

Proof:
sir rene
image

**Review by kdacudag — APPROVED** [review]
Approved
Comments: 6
Iteration: 1

**Pull request merged**